dashboard/test/controllers/regional_partners_controller_test.rb

```diff
@@ -85,4 +85,56 @@ def self.test_workshop_admin_only(method, action, params = nil)
     assert @regional_partner.program_managers.exists?(teacher.id)
     assert teacher.program_manager?
   end
+
+  test 'add mapping updates regional partner mapping' do
+    sign_in @workshop_admin
+    post :add_mapping, params: {id: @regional_partner.id, region: '12345'}
+    assert @regional_partner.mappings.length == 1
+  end
+
+  test 'replace mappings fails on invalid file' do
+    sign_in @workshop_admin
+    post :replace_mappings, params: {id: @regional_partner.id, regions: {path: 'invalid_path.csv'}}
+    assert_nil flash[:notice]
+    assert_equal(
+      'Replace mappings failed. Could not read file.',
+      flash[:alert]
+    )
+    assert @regional_partner.reload.mappings.empty?
+  end
+
+  test 'replace mappings on invalid mapping fails and does not delete old mapping' do
+    regional_partner_with_mapping = create :regional_partner_with_mappings
+    sign_in @workshop_admin
+    mapping = fixture_file_upload('regional_partner_mappings_invalid.csv', 'text/csv')
+    post :replace_mappings, params: {id: regional_partner_with_mapping.id, regions: mapping}
+    assert_nil flash[:notice]
+    assert_equal(
+      '<b>Replace mappings failed with 2 error(s):</b><br>ABC: Invalid region<br>123456: Invalid region',
+      flash[:upload_error]
+    )
+    refute regional_partner_with_mapping.mappings.empty?
+    assert_equal '98143', regional_partner_with_mapping.mappings.first.zip_code
+  end
+
+  test 'replace mappings fails on invalid non-unique mapping' do
+    create :regional_partner_with_mappings
+    sign_in @workshop_admin
+    mapping = fixture_file_upload('regional_partner_mappings.csv', 'text/csv')
+    post :replace_mappings, params: {id: @regional_partner.id, regions: mapping}
+    assert_nil flash[:notice]
+    assert_equal(
+      '<b>Replace mappings failed with 1 error(s):</b><br>98143: This region belongs to another partner',
+      flash[:upload_error]
+    )
+    assert @regional_partner.reload.mappings.empty?
+  end
+
+  test 'replace mappings succeeds on valid mapping' do
+    sign_in @workshop_admin
```

Comment:
Not something you introduced, but it looks like every test in here signs in the workshop admin -- could that be moved into the setup?

Reply:
there's actually a test at the top that doesn't sign in the admin, so will leave as is

Reply:
ooh I see 👍